src/uct/ib/mlx5/ib_mlx5.inl

```diff
@@ -388,24 +404,19 @@ uct_ib_mlx5_post_send(uct_ib_mlx5_txwq_t *wq,
     ucs_memory_bus_store_fence();
 
     /* Set up copy pointers */
-    dst = wq->bf->reg.ptr;
+    dst = wq->reg->addr.ptr;
     src = ctrl;
 
     ucs_assert(wqe_size <= UCT_IB_MLX5_BF_REG_SIZE);
     ucs_assert(num_bb <= UCT_IB_MLX5_MAX_BB);
-    if (ucs_likely(wq->bf->enable_bf)) {
-        /* BF copy */
-        for (n = 0; n < num_bb; ++n) {
-            uct_ib_mlx5_bf_copy_bb(dst, src);
-            dst += MLX5_SEND_WQE_BB;
-            src += MLX5_SEND_WQE_BB;
-            if (ucs_unlikely(src == wq->qend)) {
-                src = wq->qstart;
-            }
-        }
+    if (ucs_likely(wq->reg->mode == UCT_IB_MLX5_MMIO_MODE_BF_POST)) {
+        src = uct_ib_mlx5_bf_copy(dst, src, num_bb, wq);
+    } else if (wq->reg->mode == UCT_IB_MLX5_MMIO_MODE_BF_FLUSH) {
+        src = uct_ib_mlx5_bf_copy(dst, src, num_bb, wq);
+        ucs_memory_bus_store_fence();
```

Comment:
Do you really need bus fence here. I think on ARM you get away (in a performance portable way) with DMB ST which corresponds to CPU fence. I'm not convinced you need barrier here at all. I know you have it verbs but I think you allocation of UAR is a bit different ? Let's chat about this offline.

Reply:
we need an instruction which flushes WC memory, to ensure another thread will not overwrite it, and currently for ARM this is bus fence. if there is something more fine-grained we can define a new macro. Should we use DMB ST for arm?